### ISSUE TITLE
🐛 Bug: Gradient Disruption in loss_ob Calculation

### ISSUE THREAD

**Issue by th05785775-ux** [Feliciaxyao/NavMorph#16]
### 📝 Description
The calculated loss term, **loss_ob**, fails to contribute to the backpropagation process because the intermediate variable im_obser is constructed using the sequence of operations **.detach().cpu().numpy()**. This sequence severs the computational graph link between the loss and the input tensor (im_observation), rendering the loss term ineffective for training. This issue appears similar to the general gradient tracking concerns discussed in #15 
### 💻 Code Snippet
https://github.com/Feliciaxyao/NavMorph/blob/7df17d05a9212f0fc1d5bb0e53d43dafaee78bba/vlnce_baselines/ss_trainer_ETP.py#L1496
### 🔬 Root Cause: Non-Differentiable Computation
The root cause is twofold:
**Non-Differentiable Function:** The fastdtw function computes the DTW distance, but it does not provide gradients with respect to its inputs (im_segment and real_segment).
**Explicit Graph Disruption:** The use of .detach().cpu().numpy() for constructing im_obser explicitly converts the tensor into a NumPy array, removing it from PyTorch's Autograd system before being fed to the non-differentiable fastdtw.

When ob_dtw_distance is wrapped back into a PyTorch Tensor, it becomes a leaf tensor (no grad_fn), and the gradient flow stops there, preventing parameter updates based on loss_ob.
### 💡 Proposed Solution
To incorporate a DTW-based loss for backpropagation, a differentiable solution is required:

**Implement/Find Differentiable DTW:** Replace fastdtw with a PyTorch-native implementation of DTW (or a similar, differentiable time-series alignment method).
**Custom Autograd Function:** If no native implementation exists, define a custom torch.autograd.Function where the forward pass uses the external library, and the backward pass explicitly defines and computes the required gradients (e.g., using subgradient information or approximations).

**Comment by Feliciaxyao** [comment]
> [NavMorph/vlnce_baselines/ss_trainer_ETP.py](https://github.com/Feliciaxyao/NavMorph/blob/ae98a519cc35b526f1c872480674f9d1fc81ea63/vlnce_baselines/ss_trainer_ETP.py#L1485)
> 
> Line 1485 in [ae98a51](/Feliciaxyao/NavMorph/commit/ae98a519cc35b526f1c872480674f9d1fc81ea63)
> 
>  real_seg_tensor = real_obser_tensor[stepk - 4: stepk - 3]  # (1, D) 
> 
> Thank you for your answer, but I cannot see anywhere where this variable "real_obser_tensor" is initialized.

Thank you for pointing this out again — you are absolutely right.
The previous commit introduced a modification to the loss implementation, and during refactoring the initialization of real_obser_tensor was unintentionally omitted.

We have now added the missing initialization. The correct initialization can be found [here](https://github.com/Feliciaxyao/NavMorph/blob/7fb0f4aa44ff3e042335d8de0d1b2197c6f1f657/vlnce_baselines/ss_trainer_ETP.py#L1375)

We apologize for the confusion caused by the missing line — you should now be able to run the code without encountering this variable issue.
